src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -218,12 +218,13 @@ def extract_trusted_facts(request_data, facts)
       end
 
       def get_facts_from_pdb(nodename, environment)
+        env = environment || 'production'
         pdb_terminus = Puppet::Node::Facts::Puppetdb.new
         request = Puppet::Indirector::Request.new(pdb_terminus.class.name,
                                                   :find,
                                                   nodename,
                                                   nil,
-                                                  :environment => environment)
+                                                  :environment => env)
```

Comment:
Alternatively we can pass an options hash in here that just doesn't contain the `environment` key. In that case the `Indirector` will use whatever is the value of `Puppet.lookup(:current_environment)`. It might be possible that that value isn't "production"? I assume `:current_environment` is the environment of both the master and the agent who we're compiling the catalog for, but that other code hasn't had the chance of changing the  environment yet?

Reply:
Also, we should probably update the environment option that we pass into save methods with whatever node resolves to (as is it will be `:current_environment`).